### PR TITLE
chore: bump libs & fix libcamera, fix upstream nixpkgs

### LINKED ISSUE #19: [24.05] Build
Hey, first of all thanks for the work already done 🙏🏻

I'm currently getting the following issue - I'm trying to understand:
```
warning: The interpretation of store paths arguments ending in `.drv` recently changed. If this command is now failing try again with '/nix/store/mfxpz68z7s2vdmxmicn7cgskir3r0nym-linux-config-6.6.28.drv^*'
@nix { "action": "setPhase", "phase": "unpackPhase" }
Running phase: unpackPhase
unpacking source archive /nix/store/hdh0igwg2rj6zk6xmf8kdn7kzrv9i91h-source
source root is source
@nix { "action": "setPhase", "phase": "patchPhase" }
Running phase: patchPhase
applying patch /nix/store/mnlqwys9dxznjcvq4w3s0435s9ql8qv8-bridge-stp-helper.patch
patching file net/bridge/br_private.h
Hunk #1 succeeded at 51 with fuzz 2 (offset 12 lines).
applying patch /nix/store/mbz5gln1q4swlzmqs2399br98qjif9kn-request-key-helper.patch
patching file security/keys/request_key.c
Hunk #1 succeeded at 117 (offset 3 lines).
applying patch /nix/store/831w9jjwfspn3wc2h3mlks980v3zm2qz-6de0bb51929cd3ad4fa27b2a421a2af12e6468f5.patch
patching file drivers/gpu/drm/rp1/rp1-dpi/Kconfig
Reversed (or previously applied) patch detected!  Assume -R? [n]
Apply anyway? [n]
Skipping patch.
1 out of 1 hunk ignored -- saving rejects to file drivers/gpu/drm/rp1/rp1-dpi/Kconfig.rej
patching file drivers/gpu/drm/rp1/rp1-dsi/Kconfig
Reversed (or previously applied) patch detected!  Assume -R? [n]
Apply anyway? [n]
Skipping patch.
1 out of 1 hunk ignored -- saving rejects to file drivers/gpu/drm/rp1/rp1-dsi/Kconfig.rej
patching file drivers/gpu/drm/rp1/rp1-vec/Kconfig
Reversed (or previously applied) patch detected!  Assume -R? [n]
Apply anyway? [n]
Skipping patch.
1 out of 1 hunk ignored -- saving rejects to file drivers/gpu/drm/rp1/rp1-vec/Kconfig.rej
```

This happens when switching from 23.11 to 24.05, I'm not sure what's the issue here as the repo does not apply custom-patches?

Would you be interesting to make this project part of https://github.com/nix-community/?
What are your thoughts on any kind of GitHub-CI and automated builds/push to a binary-cache?

### PR DESCRIPTION
**Pull request by adminy**
things tested and working:
- [x] boots on pi 5 b
- [x] libcamera works
- [x] nixpkgs master works

things to try out:
 - ~[ ] uboot booting, see if its fixed.~ Uboot still not working
      - [ ] test sd card boot
 
IMG_6753

 - [ ] pi 4 test.

This should fix #19 
Untested but maybe also #16 

### Serial error
```bash
DRAM:  1016 MiB (effective 8 GiB)
RPI 5 Model B (0xd04170)
Core:  22 devices, 11 uclasses, devicetree: board
MMC:   mmc@fff000: 0, mmc@1100000: 1
Loading Environment from FAT... Unable to read "uboot.env" from mmc0:1... 
In:    serial,usbkbd
Out:   serial,vidconsole
Err:   serial,vidconsole
mbox: Header response code invalid
bcm2835: Could not query MAC address
Net:   No ethernet found.

starting USB...
No working controllers found
Hit any key to stop autoboot:  0 
No EFI system partition
No EFI system partition
Failed to persist EFI variables
No EFI system partition
Failed to persist EFI variables
No EFI system partition
Failed to persist EFI variables
** Booting bootflow '<NULL>' with efi_mgr
Loading Boot0000 'mmc 0' failed
EFI boot manager: Cannot load any image
Boot failed (err=-14)
** Booting bootflow 'mmc@fff000.bootdev.part_2' with extlinux
------------------------------------------------------------
1:      NixOS - Default
Enter choice: 1 
1:      NixOS - Default
Retrieving file: /boot/extlinux/../nixos/4ljfp9dpyfv3r0pw1b8nwwxysc2c59fx-linux-rpi-6.6.34-Image
Retrieving file: /boot/extlinux/../nixos/iphdbw2np9dsacx7i37xbazqkczjsiaz-initrd-linux-rpi-6.6.34-initrd
append: init=/nix/store/4cpd34134wcaxqy4hj0dchkq020ccnb3-nixos-system-box_8-24.11/init console=serial0,115200n8 console=tty1 loglevel=7
Moving Image from 0x80000 to 0x200000, end=27a0000
ERROR: RD image overlaps OS image (OS=200000..27a0000)
Boot failed (err=-14)
No working controllers found
No ethernet found.
No ethernet found.
U-Boot> 

```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -27,9 +27,10 @@ complete example.
 {
   description = "raspberry-pi-nix example";
   nixConfig = {
-    extra-substituters = [ "https://raspberry-pi-nix.cachix.org" ];
+    extra-substituters = [ "https://raspberry-pi-nix.cachix.org" https://adminy.cachix.org ];
```
Comment (SebTM) [inline README.md:30]: Is this needed upstream? We should commit on one binary-cache I would argue

Comment (adminy) [reply README.md]: No problem. I removed it. for anyone wanting a cached kernel build, you can now get it from the previous commit :)

**Comment by SebTM** [comment]
Thanks for the fast update (and providing binary-cache to test) - the changes itself look good to me but I encountered issues compared to building against master. When booting the pi networking is somehow broken - I haven't debugged further - just took that photo when plugging a monitor for now:

photo_2024-06-22_20-32-40

Normally the pi is configured to join by wifi after boot.

**Comment by adminy** [comment]
I noticed some wifi issues also. could be that the latest wifi firmware is unstable. I switched to 5ghz and the issue still persists but I see no warnings. something's not right with the firmware.

File: rpi/config.nix [hunk 2]
```diff
@@ -16,8 +16,7 @@ let
           (lib.filterAttrs (k: v: v.enable) x);
       render-dt-overlay = { overlay, args }:
         "dtoverlay=" + overlay + "\n"
-        + lib.strings.concatMapStringsSep "\n" render-dt-param args + "\n"
-        + "dtoverlay=";
```
Comment (tstat) [inline rpi/config.nix:20]: Why was this removed? It was there for a reason: it ends the `dtoverlay` parameter scope.

File: rpi/default.nix [hunk 3]
```diff
@@ -301,9 +312,12 @@ in
         "reset-raspberrypi" # required for vl805 firmware to load
       ];
       # This pin is not necessary, it would be fine to replace it with
-      # `pkgs.rpi-kernels.latest.kernel`. It is helpful to ensure
+      # `kernel`. It is helpful to ensure
       # cache hits for kernel builds though.
-      kernelPackages = pkgs.linuxPackagesFor pkgs.rpi-kernels.latest.kernel;
+      kernelPackages = pkgs.linuxPackagesFor (kernel.override {
+        # Some patches cannot be applied because they are already upstream.
+        ignoreConfigErrors = true;
```
Comment (tstat) [inline rpi/default.nix:319]: Can you motivate this?

**Comment by SebTM** [comment]
I see, I would suggest (not a maintainer of the project) to maybe split up the PR with the "safe" / "stable" and the wireless update in separate then maybe?

**Review by tstat — COMMENTED** [review]
Thanks for this PR! I put this stuff on github in the first place because I want to collaborate on this work with others, so I welcome this PR.

That being said, I think there are a number of uncontroversial changes in here, but they are mixed in with other unrelated changes that require more discussion. 

It would be easier to consider PRs that have a more narrow focus. So, I would recommend closing this PR and opening a number of PRs, each with a specific focus (e.g. dependency bumps, adding defconfig to kernel, etc).

File: overlays/default.nix [hunk 4]
```diff
@@ -34,17 +70,12 @@ let
   rpi-kernels = builtins.foldl' (b: a: b // rpi-kernel a) { };
 in
 {
-  # disable firmware compression so that brcm firmware can be found at
-  # the path expected by raspberry pi firmware/device tree
-  compressFirmwareXz = x: x;
-  compressFirmwareZstd = x: x;
-
```
[the remark this answers is not in the document]
Comment (tstat) [reply overlays/default.nix:41]: Dropping these overrides would cause network issues, so dropping these from the overlay is probably the source of the network issues @SebTM mentions above.

File: flake.nix [hunk 5]
```diff
@@ -2,38 +2,38 @@
   description = "raspberry-pi nixos configuration";
 
   inputs = {
-    nixpkgs.url = "github:NixOS/nixpkgs/9a9960b98418f8c385f52de3b09a63f9c561427a";
+    nixpkgs.url = "github:NixOS/nixpkgs/9be7393d7204ba14ead6ec4f0381cc91bfb2aa24"; # 2024-06-18
     u-boot-src = {
       flake = false;
-      url = "https://ftp.denx.de/pub/u-boot/u-boot-2024.04.tar.bz2";
+      url = "https://ftp.denx.de/pub/u-boot/u-boot-2024.07-rc4.tar.bz2";
```
[the remark this answers is not in the document]
Comment (adminy) [reply flake.nix:8]: no reason, I was hoping it might have the nvme boot. But the actual issue seems to be that u-boot can't load the kernel and initrd because its too large

File: flake.nix [hunk 6]
```diff
@@ -2,38 +2,38 @@
...
     };
     rpi-linux-6_6-src = {
       flake = false;
-      url = "github:raspberrypi/linux/stable_20240423";
+      url = "github:raspberrypi/linux/da87f91ad8450ccc5274cd7b6ba8d823b396c96f"; # 2024-06-17
```
[the remark this answers is not in the document]
Comment (adminy) [reply flake.nix:12]: libcamera won't work in that case I think. Also there was one more compilation issue with it that was fixed specifically at this hash. But lets try see if latest stable compiles.

File: overlays/default.nix (hunk 4 above)
[the remark this answers is not in the document]
Comment (adminy) [reply overlays/default.nix:41]: ooooo. Okay.

File: overlays/libcamera.nix [hunk 7]
```diff
@@ -22,9 +22,30 @@ final: prev:
...
+      "-Dv4l2=true"
+      "-Dgstreamer=enabled"
+      "-Dtest=false"
+      "-Dlc-compliance=disabled"
+      "-Dcam=disabled"
+      "-Dqcam=disabled"
+      "-Ddocumentation=enabled"
+      "-Dpycamera=enabled"
+    ];
+
```
[the remark this answers is not in the document]
Comment (adminy) [reply overlays/libcamera.nix:49]: libcamera version 2 now requires these flags to build. it was in their readme.

File: rpi/config.nix (hunk 2 above)
Comment (adminy) [reply rpi/config.nix:20]: I was getting an extra empty dtoverlay= in the generated config.

File: rpi/default.nix (hunk 3 above)
Comment (adminy) [reply rpi/default.nix:319]: Latest kernel has patches that could not be overwritten, so rather than not using nixpkgs rpi-kernels nix configuration, I just said, ignore failed patch applies.

File: flake.nix (hunk 6 above)
[the remark this answers is not in the document]
Comment (adminy) [reply flake.nix:12]: also another thing: https://github.com/raspberrypi/linux/issues/6237

File: rpi/config.nix (hunk 2 above)
Comment (tstat) [reply rpi/config.nix:20]: I see how that would look like an error, but that line is meaningful; It delimits the scope of the parameters associated with the above overlay. For example:

```
dtoverlay=<SOME_OVERLAY>
param0=val0
param1=param1
...
dtoverlay=
```
This ensures that all params listed above are interpreted as parameters to `<SOME_OVERLAY>`, and avoids ambiguity if a parameter name occurs both in an overlay and a base config.

See the [docs](https://www.raspberrypi.com/documentation/computers/configuration.html#part3.1).

specifically, this part:
Screenshot 2024-06-22 at 6 35 51 PM

File: overlays/default.nix [hunk 8]
```diff
@@ -7,25 +7,61 @@
...
+  rpi-kernel = { kernel, version, fw, wireless-fw, argsOverride ? null, board }:
     let
       new-kernel = prev.linux_rpi4.override {
         argsOverride = {
           src = kernel;
           inherit version;
           modDirVersion = version;
+          kernelPatches = [];
+          defconfig = "${board}_defconfig";
+          postFixup = "";
```
[the remark this answers is not in the document]
Comment (adminy) [reply overlays/default.nix:57]: Also maybe untangle the boot loader and wireless firmware from it. Because I don't think those are board or kernel dependent. I'll see what I can do.

File: overlays/libcamera-apps.nix [hunk 9]
```diff
@@ -1,28 +1,12 @@
...
-, libtiff
-, libpng
-, libcamera
-, libepoxy
-, boost
-, libexif
-, ninja
-}:
-
+{ libcamera-apps-src, lib, pkgs, stdenv }:
```
[the remark this answers is not in the document]
Comment (adminy) [reply overlays/libcamera-apps.nix:1]: well, those are coming from pkgs, Its mentally easier to wrap your head around the fact that callPackage just provides pkgs rather than it also does a spread, which confuses me, because we're not passing all these extra params.

File: flake.nix [hunk 10]
```diff
@@ -2,38 +2,38 @@
   description = "raspberry-pi nixos configuration";
 
   inputs = {
-    nixpkgs.url = "github:NixOS/nixpkgs/9a9960b98418f8c385f52de3b09a63f9c561427a";
+    nixpkgs.url = "github:NixOS/nixpkgs/9be7393d7204ba14ead6ec4f0381cc91bfb2aa24"; # 2024-06-18
```
[the remark this answers is not in the document]
Comment (adminy) [reply flake.nix]: done

File: overlays/libcamera.nix [hunk 11]
```diff
@@ -11,7 +11,7 @@ final: prev:
 
   libpisp = final.stdenv.mkDerivation {
     name = "libpisp";
-    version = "1.0.3";
+    version = "1.0.5";
```
[the remark this answers is not in the document]
Comment (adminy) [reply overlays/libcamera.nix]: resolved.

File: rpi/config.nix (hunk 2 above)
Comment (adminy) [reply rpi/config.nix]: undid that, thanks for the explanation!

File: overlays/default.nix (hunk 8 above)
[the remark this answers is not in the document]
Comment (adminy) [reply overlays/default.nix]: updated this to `pkgs.rpi-kernels.<VERSION>.<BOARD>`, and untangled it from the firmware. What do you think of the new code @tstat ?

[the remark this answers is not in the document]
Comment (tstat) [reply overlays/default.nix]: I'm on mobile right now, but it looks great! I'll give it a closer look when I'm at a computer later.

**Comment by tstat** [comment]
I went to build an image with this branch yesterday, but there were some syntax errors in the kernel derivation. I pushed some fixes and will test an image later today.

**Comment by adminy** [comment]
So a silly mistake happened. My lock file locked a version of the local copy I had, I thought I was making testable changes but it was just running an older version of the flake :D Thanks for your changes, much appreciated!

**Comment by adminy** [comment]
I have tested it with 6.6.34 and it works fine, with the exception of the wifi, there seems to be an issue with it being terribly slow/choppy (but no patches required). Compiling the 6.6.31 also went smoothly! No more ignore errors!
Also on 6.6.31 the wifi has no issues! So good that its left as default.

**Comment by adminy** [comment]
Screenshot 2024-06-28 at 07 57 27
Tested again, fixed wifi issues in the latest rc pi linux it seems.

**Comment by tstat** [comment]
Thanks for the contribution!
